pkg/querier/frontend/frontend_test.go

```diff
@@ -251,8 +255,75 @@ func TestFrontendCheckReady(t *testing.T) {
 	}
 }
 
-func testFrontend(t *testing.T, config Config, handler http.Handler, test func(addr string), matchMaxConcurrency bool) {
+func TestFrontend_LogsSlowQueriesFormValues_Issue3111(t *testing.T) {
+	// Create an HTTP server listening locally. This server mocks the downstream
+	// Prometheus API-compatible server.
+	downstreamListen, err := net.Listen("tcp", "localhost:0")
+	require.NoError(t, err)
+
+	downstreamServer := http.Server{
+		Handler: http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+			_, err := w.Write([]byte(responseBody + responseBody + responseBody + responseBody + responseBody))
```

Comment:
Could you comment on why are multiple `responseBody` chained together, please?

Reply:
Yes, at first I had to somehow trigger closing body and was playing with ResponseWriter which needs some data written to actually trigger request body close. 
Tested it now with short body and it also triggers this error with downstream server so it should be enough.